### DISCUSSION TITLE
Questions about MemoryPool

### DISCUSSION THREAD

**Discussion by eedalong** [facebookincubator/velox#13461] [General]
First of all, great project guys! But 2 things about MemoryPool implementation confuse me a little bit.

1.  Seems that you dont actually reuse the already allocated pages but choose to malloc new pages everytime when you call `MemoryPoolImpl::allocte`

2. why you choose to use mmap for memory pool implementation? Any reason behind this choice?

**Comment by mbasmanova** [comment]
CC: @tanjialiang @oerling @xiaoxmeng

**Comment by oerling** [comment]
Using mmap as we do gives you a precisely known resident set size and guaranteed zero fragmentation. This is needed for running inside a fixed resource envelope. Also, you know how much memory you actually are holding. The same is not true of malloc. But there is also a malloc based variant of the functions.  Of course if you use malloc then malloc keeps its free lists etc and you in general would not be performing this function. There are a couple of instances of more complex objects (consisting of multiple pieces of memory), like vectors, being recycled, e.g. VectorPool.

**Comment by eedalong** [comment]
Thx, I am clear about why you use mmap. What about memory reuse?


From what I see, your memory management behaves like this when user requests new memory allocation.

# Standard Memory 

Use `malloc` for each allocation

# MappedMemory

- Use `malloc` directly for <3KB allocation
- Use `malloc` for 3KB < bytes < largest pagerun size, but roundup to a certain pagerun size
- Use 'mmap' for beyond the largest single pagerun size allocation. 

Your memory pool use these 2 for memory allocation and free, but seems that I dont see any memory reuse here ?
